kani-compiler/src/kani_middle/reachability.rs

```diff
@@ -400,11 +401,29 @@ impl<'a, 'tcx> MirVisitor<'tcx> for MonoItemsFnCollector<'a, 'tcx> {
                 let callee_ty = func.ty(self.body, tcx);
                 let fn_ty = self.monomorphize(callee_ty);
                 if let TyKind::FnDef(def_id, substs) = *fn_ty.kind() {
-                    let instance =
+                    let instance_opt =
                         Instance::resolve(self.tcx, ParamEnv::reveal_all(), def_id, substs)
-                            .unwrap()
                             .unwrap();
-                    self.collect_instance(instance, true);
+                    match instance_opt {
+                        None => {
+                            // During the MIR stubbing transformation, we do not
+                            // force type variables in the stub's signature to
+                            // implement the same traits as those in the
+                            // original function/method. A trait mismatch shows
+                            // up here, when we try to resolve a trait method
+                            // that the type does not implement.
+                            tcx.sess.span_err(
+                                terminator.source_info.span,
+                                format!(
+                                    "Unable to resolve call to `{}`; \
+                        this is either an internal Kani bug, or there is a trait mismatch \
```

Comment:
I'd avoid saying "this is either an internal Kani bug". Of course, it's possible we have a bug here (or anywhere in the tool, especially when it comes to unstable features), but let's first make clear what the error we found was, so the user can take action on it. They can always open an issue if they can't figure it out.

Reply:
I've changed this, so if we're in the body of a function that has been stubbed it says something like:
```
error: unable to resolve call to `DoIt::do_it`; there is likely a trait mismatch between the signatures for the stub `bar` and the original function/method `foo`
```

Reply:
Okay, it now has this message:
```
error: `&&str` doesn't implement `DoIt`. The function `foo` cannot be stubbed by `bar` due to generic bounds not being met.
```
I could probably remove the `&`s if that's preferable.

Reply:
It depends where the `&` comes from. The type you print should be the type you are trying to check the bounds against. In some cases it will have `&`, but others won't.

Reply:
Cool - I updated to that behavior, so that we now print this if `T == &str`:
```
error: `&str` doesn't implement `std::cmp::PartialEq`. The function `foo` cannot be stubbed by `bar` due to generic bounds not being met.
```
but this if `T == String`:
```
error: `std::string::String` doesn't implement `DoIt`. The function `foo` cannot be stubbed by `bar` due to generic bounds not being met.
```